discord/sku.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+from typing import Optional, TYPE_CHECKING
+
+from . import utils
+from .enums import try_enum, SKUType, EntitlementType
+from .flags import SKUFlags
+
+if TYPE_CHECKING:
+    from datetime import datetime
+
+    from .guild import Guild
+    from .state import ConnectionState
+    from .types.sku import (
+        SKU as SKUPayload,
+        Entitlement as EntitlementPayload,
+    )
+
+__all__ = (
+    'SKU',
+    'Entitlement',
+)
+
+
+class SKU:
+    """Represents a premium offering as a stock-keeping unit (SKU).
+
+    .. versionadded:: 2.4
+
+    Attributes
+    -----------
+    id: :class:`int`
+        The SKU's ID.
+    type: :class:`SKUType`
+        The type of the SKU.
+    application_id: :class:`int`
+        The ID of the application that the SKU belongs to.
+    name: :class:`str`
+        The consumer-facing name of the premium offering.
+    slug: :class:`str`
+        A system-generated URL slug based on the SKU name.
+    """
+
+    __slots__ = (
+        '_state',
+        'id',
+        'type',
+        'application_id',
+        'name',
+        'slug',
+        '_flags',
+    )
+
+    def __init__(self, *, state: ConnectionState, data: SKUPayload):
+        self._state: ConnectionState = state
+        self.id: int = int(data['id'])
+        self.type: SKUType = try_enum(SKUType, data['type'])
+        self.application_id: int = int(data['application_id'])
+        self.name: str = data['name']
+        self.slug: str = data['slug']
+        self._flags: int = data['flags']
+
+    def __repr__(self) -> str:
+        return f'<SKU id={self.id} name={self.name!r} slug={self.slug!r}>'
+
+    @property
+    def flags(self) -> SKUFlags:
+        """Returns the flags of the SKU."""
+        return SKUFlags._from_value(self._flags)
+
+
+class Entitlement:
```

Comment:
This type needs a few things for ergonomics:

1. A `user` property similar to `guild`.
2. Probably an `sku` property that gives either an `SKU` or `PartialSKU` type?
3. An `is_valid()` method to check if the entitlement is still valid
4. The `delete()` method from earlier.

Reply:
2.) I can't create an SKU, hence I'm missing crucial information. A PartialSKU would just store the ID. Not sure if it worth it to create such class.
3.) What is valid? Just checking the expiry, then `is_expired()` might be better. A future proof `valid` method could include the consumed attribute, if Discord adds support for one-time purchases or uses the same api for the new server shop feature